src/games/strategy/triplea/ui/MapData.java

```diff
@@ -700,6 +701,50 @@ private Rectangle getBoundingRectWithTranslate(final List<Polygon> polys, final
   }
 
   /**
+   * Find the best rectangle inside the territory to place the name in. Finds the rectangle
+   * that can fit the name, that is the closest to the vertical center, and has a large width at
+   * that location. If there isn't any rectangles that can fit the name then default back to the
+   * bounding rectangle.
+   */
+  public Rectangle getBestTerritoryNameRect(final Territory t, final FontMetrics fm) {
+    final Rectangle territoryBounds = getBoundingRect(t);
+    Rectangle maxRectangle = territoryBounds;
+    final int minX = territoryBounds.x;
+    final int minY = territoryBounds.y;
+    final int maxX = minX + territoryBounds.width;
+    final int maxY = minY + territoryBounds.height;
+    final int centerY = minY + territoryBounds.height / 2;
+    final List<Polygon> polygons = m_polys.get(t.getName());
+    final int minWidth = fm.stringWidth(t.getName());
+    final int increment = fm.getAscent();
+    int maxScore = 0;
+    for (int startX = minX; startX < maxX - increment; startX += increment) {
+      for (int startY = minY; startY < maxY - increment; startY += increment) {
+        for (int endX = maxX; endX > startX; endX -= increment) {
+          final Rectangle current = new Rectangle(startX, startY, endX - startX, increment);
+          final int verticalFromEdge = territoryBounds.height / 2 - Math.abs(centerY - (startY + increment));
+          final int score = verticalFromEdge * current.width;
+          if (current.width > minWidth && score > maxScore) {
+            boolean isContained = false;
+            for (final Polygon polygon : polygons) {
+              if (polygon.contains(current)) {
```

Comment:
What is that max value. Please indicate it in the code so it's clear which ranges of values we will have in the Math.abs function.

If verticalFromEdge is at the max when Y is half way, then start your loop there. There are two linear functions, and an absolute function, which you should be able to treat as linear. Given it's all linear, you should be able to start with the max value and get this down to `O(n)`